src/cluster_resources.rs

```diff
@@ -0,0 +1,393 @@
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    /// * `resource` - A resource to add to the cluster
+    ///
+    /// # Errors
+    ///
+    /// If the labels of the given resource are not set properly then an `Error::MissingLabel` or
+    /// `Error::UnexpectedLabelContent` is returned. The expected labels are:
+    /// * `app.kubernetes.io/instance = <cluster.name>`
+    /// * `app.kubernetes.io/managed-by = <app_name>-operator`
+    /// * `app.kubernetes.io/name = <app_name>`
+    pub async fn add<T: ClusterResource>(
+        &mut self,
+        client: &Client,
+        resource: &T,
+    ) -> OperatorResult<T> {
+        ClusterResources::check_label(resource.labels(), APP_INSTANCE_LABEL, &self.app_instance)?;
+        ClusterResources::check_label(resource.labels(), APP_MANAGED_BY_LABEL, &self.manager)?;
+        ClusterResources::check_label(resource.labels(), APP_NAME_LABEL, &self.app_name)?;
+
+        let patched_resource = client
+            .apply_patch(&self.manager, resource, resource)
+            .await?;
+
+        self.resources.insert((&patched_resource).into());
+
+        Ok(patched_resource)
+    }
+
+    /// Checks that the given `labels` contain the given `label` with the given `expected_content`.
+    ///
+    /// # Arguments
+    ///
+    /// * `labels` - The labels to check
+    /// * `label` - The expected label
+    /// * `expected_content` - The expected content of the label
+    ///
+    /// # Errors
+    ///
+    /// If `labels` does not contain `label` then an `Error::MissingLabel` is returned.
+    ///
+    /// If `labels` contains the given `label` but not with the `expected_content` then an
+    /// `Error::UnexpectedLabelContent` is returned
+    fn check_label(
+        labels: &BTreeMap<String, String>,
+        label: &'static str,
+        expected_content: &str,
+    ) -> OperatorResult<()> {
+        if let Some(actual_content) = labels.get(label) {
+            if expected_content == actual_content {
+                Ok(())
+            } else {
+                Err(Error::UnexpectedLabelContent {
+                    label,
+                    expected_content: expected_content.into(),
+                    actual_content: actual_content.into(),
+                })
+            }
+        } else {
+            Err(Error::MissingLabel { label })
+        }
+    }
+
+    /// Finalizes the cluster creation.
+    ///
+    /// All orphaned resources, i.e. resources which are labelled as if they belong to this cluster
+    /// instance but were not added to the cluster resources, are deleted.
+    ///
+    /// The following resource labels are compared:
+    /// * `app.kubernetes.io/instance`
+    /// * `app.kubernetes.io/managed-by`
+    /// * `app.kubernetes.io/name`
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - The client which is used to access Kubernetes
+    pub async fn finalize(self, client: &Client) -> OperatorResult<()> {
```

Comment:
Just a comment, not sure if this happened to others as well: Before actually reading the fn I kept thinking this was something to do with finalizers and was surprised that this is not the case. In any other context the name fits I think, but in k8s it migth already be overloaded?

Reply:
I see three options to reduce the confusion:

1. Use a synonym of `finalize`, like `close` or `complete`, to underline that this is the terminal operation of `ClusterResources`.
2. Use a name describing what the function is actually doing, like `delete_orphaned_resources` or `delete_remaining_cluster_resources`.
3. Convert this function into a `Drop` trait so that it is implicitly called. This could be difficult to achieve with the asynchronous context and it increases the magic behind the scenes.

Which option do you prefer?

Reply:
I think I'm in favor of 2.

Creating resources is done upon adding them to the struct, not when calling this fn, correct?

Reply:
I renamed the function to `delete_orphaned_resources`.

> Creating resources is done upon adding them to the struct, not when calling this fn, correct?

Correct. This is necessary because some operators use the data from the patched resources to create new resources, e.g. when creating discovery maps.